### PR TITLE
Feat/playwright smoke 332

### LINKED ISSUE #332: Add a Playwright smoke suite for the built dashboard
> 🧑‍💻 **Intermediate Issue** — a complex task spanning multiple modules, with real design decisions to own.
> **Time:** ~25 hours · **Prerequisites:** comfortable navigating this repo (a completed [beginner issue](https://github.com/hiero-hackers/analytics/issues?q=is%3Aissue+state%3Aopen+no%3Aassignee+label%3Abeginner) is the usual route; demonstrated CI/CD proficiency substitutes for workflow-focused issues).
> We expect more than "it works": maintainable code that fits the existing architecture.

## The task

**Problem:** The vitest suite renders components against TypeScript fixtures — it never exercises the built `web/dist`, the real `fetch` path, or the manifest-driven boot sequence. That last mile is exactly where the system is most fragile: the dashboard renders whatever the manifest lists, so a Python-side change that emits a subtly wrong manifest (renamed path, missing section file) ships green through both `uv run pytest` (which checks the emit side) and `npm test` (which checks components against its own fixtures), and breaks only in production. Nothing today loads the built app in a real browser against files on disk.

**What done looks like:** A Playwright smoke suite that: builds `web/`, serves `dist/` with a minimal static `data/api/v1` fixture placed under the same root (the app resolves its API from `BASE_URL` — see `API_ROOT` in `web/src/api.ts` — so no proxying is needed), and asserts in Chromium that the app boots with **zero console errors**, every macro tab in the fixture manifest renders, at least one section table shows rows, and the chart lightbox opens and closes. The suite runs in CI on every PR in a couple of minutes, fully offline — it must never touch the network or the live site.

**Modules involved / constraints:**
- `web/` — Playwright as a devDependency, a `playwright.config.ts`, an `e2e/` directory, and npm scripts; Chromium only (this is a smoke layer, not a browser-matrix).
- The design decision to own is **where the API fixture comes from**. A hand-written JSON blob under `web/e2e/fixtures/` will drift from the real manifest shape silently — the exact failure mode this suite exists to catch. Recommendation: generate the static JSON from the existing typed fixtures in `web/src/test/fixtures.ts` via a small script at test-setup time, so the unit suite and the e2e suite share one source of truth and the TypeScript compiler keeps the fixture honest against `api.ts` types. If you see a better route (e.g. a checked-in golden emitted by the Python exporter from tiny input CSVs — stronger, but couples the suites across languages), argue it in your approach comment.
- CI wiring: extend `.github/workflows/test.yml` with a job (or add a workflow) that caches npm and the Playwright browser download; keep total added CI time under ~3 minutes. Mind the repo's top-level workflow permissions conventions (#314) and pin action versions by SHA like the existing workflows do.
- Keep scope disciplined: this issue is boot + render + one interaction. Visual regression, multi-browser, and accessibility audits are follow-ups, not scope creep to include.

## How to work on this

1. **Claim it:** comment `/assign` and wait to be assigned — unassigned PRs are closed automatically.
2. **Get a plan:** once assigned, comment `@coderabbitai plan` for a draft plan, then do your own investigation — [docs/architecture.md](https://github.com/hiero-hackers/analytics/blob/main/docs/architecture.md) maps the layers and their rules.
3. **Propose your approach as a comment before coding.** A paragraph is enough; early feedback here routinely saves days of rework.

**🤖 AI:** tools are welcome; verified work is required — you can explain every line and defend every design choice. See the [AI policy](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#ai-policy). Fully automated bot PRs are closed.

**Worth knowing about this repo before you design:**

- The layer rules in [docs/architecture.md](https://github.com/hiero-hackers/analytics/blob/main/docs/architecture.md) are strict — review will hold your solution to them.
- Tests mirror src (`tests/<pkg>/test_<module>.py`), and the output-contract test pins the pipeline output surface — if your change adds or renames outputs, update the contract deliberately.

**Before opening your PR:**

- [ ] I proposed my approach on this issue and incorporated any feedback
- [ ] The solution fits the existing architecture and layer rules, and is clear enough for others to debug without me
- [ ] Tests cover the happy path, edge cases, and error handling ([testing guide](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#testing))
- [ ] I reviewed my own diff line by line; scope is limited to this issue
- [ ] Workflow checks pass — CI green, signed commits, linked issue

**Stuck?** Comment here with what you've tried — see [getting help](https://github.com/hiero-hackers/analytics/blob/main/CONTRIBUTING.md#getting-help).

### PR DESCRIPTION
**Pull request by Mounil2005**
**Description:**

Add a Playwright smoke suite that loads the built dashboard in Chromium against files on disk. The vitest suite stubs fetch and renders components against fixture objects, so the built bundle and the manifest-driven boot are never exercised.

- Add a Chromium smoke suite covering boot, macro tabs, section rows and the chart lightbox
- Generate the e2e API fixture from src/test/fixtures.ts so both suites share one source of truth
- Move stubApi into its own module, freeing the fixture data of vitest imports
- Serve the staged site from a small static server that behaves like GitHub Pages
- Scope the vitest include to src/ so it stops collecting Playwright specs
- Add tsconfig.e2e.json so tsc -b typechecks e2e/
- Pin the three constants the frontend mirrors from Python, so drift fails CI
- Run the suite in a CI job parallel to the existing web job

**Related issue(s):**
Fixes #332

**Notes for reviewer:**

- Two config gaps had to be closed first. Vitest had no include, so its default pattern collected the Playwright specs and broke npm test. And tsc -b covered only src and vite.config.ts, so e2e/ was never typechecked, a deliberate type error there exited 0. The second matters: generating the fixture from typed data is only worth anything if TypeScript actually checks it.

- Not vite preview for the server: it inherits server.proxy, forwards /data and /charts to the dev server on :8642, and answers 502. It also falls back to index.html for unknown paths, hiding 404s. Pages is a plain static host, so e2e/serve.ts is one too.

- I skipped the Python golden, test_output_contract.py and test_data_api.py already pin the emit side. What no test covered until now is the browser: the bundle booting, base-relative URLs, and the CSP in index.html, which jsdom doesn't enforce. That is what this suite adds.

- The second commit closes a gap this work exposed. Three values are declared in Python and restated in the frontend — ColumnFormat in api.ts, FORMULA_PREFIXES in safety.ts, and the :root heat ramp in app.css, each held in step by a source comment and nothing else. Adding a format in Python left pytest, npm test and the browser suite all green, because the e2e fixture is generated from the TypeScript side and can only contain formats the union already allows. The prefixes matter most: drift there silently unprotects one of the two CSV download paths against formula injection.

- stubApi moves module unchanged, it carries #368's overrides implementation verbatim.

- Every guard was checked to fail when it should, not assumed: a section with no route fails staging, an inline script fails on the real CSP violation, an external image with img-src relaxed fails the offline guard, and drifting any of the three mirrored constants fails with the exact diff. The ramp check is scoped to :root, since the dark-mode block redefines those properties on purpose, verified by confirming a dark-ramp change is correctly ignored.

- Runs in under 10s locally. npm test unchanged at 54 tests.

### CODE REVIEW

**Comment by exploreriii** [comment]
Hi @Mounil2005 is this PR ready to review yet?

**Comment by Mounil2005** [comment]
> Hi @Mounil2005 is this PR ready to review yet?

Yes, pls!

File: web/e2e/serve.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+await buildSite();
+
+const server = createServer((request, response) => {
+  const { pathname } = new URL(request.url ?? '/', `http://${HOST}:${PORT}`);
+  const requested = decodeURIComponent(pathname).replace(/^\/+/, '');
+  const file = resolve(SITE_DIR, requested === '' ? 'index.html' : requested);
+
+  // Nothing outside the staged site is servable, whatever the path traversal.
+  if (file !== SITE_DIR && !file.startsWith(SITE_DIR + sep)) {
```
Comment (phillip-nyinomujuni) [inline web/e2e/serve.ts:42]: nice, appending `sep` here actually matters — without it a sibling dir like `SITE_DIR-evil` would sneak past the prefix check.

File: web/e2e/dashboard.spec.ts [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+ */
+async function watchPage(page: Page, baseUrl: string): Promise<Watch> {
+  const problems: string[] = [];
+  const external: string[] = [];
+  // Compared by origin rather than string prefix: `http://127.0.0.1:4173@evil.test/`
+  // starts with the base URL but resolves to `evil.test`, so a prefix test would
+  // wave through the very thing this guard exists to catch. Anything unparseable
+  // counts as off-origin too.
+  const baseOrigin = new URL(baseUrl).origin;
+  const isLocal = (url: string) => {
```
Comment (phillip-nyinomujuni) [inline web/e2e/dashboard.spec.ts:40]: good call using `.origin` instead of a string prefix, `http://base@evil.test/` would've slipped through otherwise.

File: tests/contracts/test_frontend_constants.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+# embedded quote of the same kind, so this needs no escape handling.
+_STRING = re.compile("'[^']*'|\"[^\"]*\"")
+
+COLUMN_FORMAT_UNION = re.compile(r"export type ColumnFormat\s*=(?P<body>[^;]*);")
+FORMULA_PREFIX_ARRAY = re.compile(r"const FORMULA_PREFIXES\s*=\s*\[(?P<body>[^\]]*)\]")
+# The ramp is spread over one custom property per bucket, not a single list, and
+# it is declared twice: once in `:root` and again, deliberately inverted, in the
+# dark-mode block. Only the light one has a Python counterpart, so the search is
+# scoped to `:root` rather than taking whichever match comes first.
+ROOT_BLOCK = re.compile(r"^:root\s*\{(?P<body>.*?)^\}", re.MULTILINE | re.DOTALL)
```
Comment (phillip-nyinomujuni) [inline tests/contracts/test_frontend_constants.py:45]: Does this still match if `:root` ever ends up nested inside a `@layer` block instead of starting at column 0? The `^\{` anchor assumes it's at the start of a line. Not sure if that's actually possible in this codebase's CSS structure — did you test it against a nested case, or is `:root` guaranteed to always be top-level here?

File: web/package.json [hunk 4]
```diff
@@ -10,7 +10,9 @@
     "format": "prettier --write .",
     "format:check": "prettier --check .",
     "test": "vitest run",
-    "preview": "vite preview"
+    "preview": "vite preview",
+    "e2e:serve": "npm run build && node e2e/serve.ts",
+    "test:e2e": "playwright test"
```
Comment (phillip-nyinomujuni) [inline web/package.json:15]: Is `e2e:serve` actually called anywhere? I don't see it in the CI workflow or `playwright.config.ts` looks like the config drives its own build through `webServer`. Is this a manual debug script, or did it get orphaned?

**Review by phillip-nyinomujuni — COMMENTED** [review]
No blockers from me. Two open questions above (the CSS regex assumption and the maybe-unused script) otherwise looks solid.

File: web/e2e/serve.ts (hunk 1 above)
Comment (Mounil2005) [reply web/e2e/serve.ts:50]: Thanks. That's exactly why it's there: SITE_DIR-evil starts with SITE_DIR. I re-tested it while rebasing. An encoded ..%2f..%2f gets a 403, and plain .. paths are normalised away and 404. That also turned up a separate bug: a malformed % escape made decodeURIComponent throw and crashed the server. It now returns a 400.

File: web/e2e/dashboard.spec.ts (hunk 2 above)
Comment (Mounil2005) [reply web/e2e/dashboard.spec.ts:40]: Thanks. http://127.0.0.1:4173.evil.test/ also starts with the base URL, so the origin check covers that one too. Anything that doesn't parse counts as off-origin.

File: tests/contracts/test_frontend_constants.py (hunk 3 above)
Comment (Mounil2005) [reply tests/contracts/test_frontend_constants.py]: Good question, and no, I hadn't tested it. I have now, and it was more fragile than it looked. Wrapping :root in an @layer failed the test, and so did adding a second top-level :root. Both failed loudly, so neither could pass wrongly, but the second was a false alarm. The ^ was standing in for "not inside the dark @media".

In this file :root is top level today and the dark one is indented inside @media, so it worked, but only by layout. I replaced it with a small brace-matching scan that keeps every :root rule no @media, @supports or @container wraps. @layer is allowed since it only orders the cascade. It has its own cases for nesting, a second block, the @layer a, b; statement at the top of app.css, comments containing braces, and unbalanced input. I broke the scan three different ways and each was caught.

File: web/package.json (hunk 4 above)
Comment (Mounil2005) [reply web/package.json:15]: It's used. webServer.command in playwright.config.ts is npm run e2e:serve, so the config doesn't drive its own build. It delegates to this script, which builds, stages the fixture site and serves it. That's also why CI only calls npm run test:e2e. You can run it by hand to open the fixture dashboard in a browser.

File: web/e2e/serve.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
+/**
```
Comment (exploreriii) [inline web/e2e/serve.ts:1]: i think we can delete this file 
```
// vite.config.ts
appType: 'mpa',              
preview: { proxy: {} },      
```

File: tests/contracts/test_frontend_constants.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
+"""Cross-language contract: constants Python declares and the frontend restates.
```
Comment (exploreriii) [inline tests/contracts/test_frontend_constants.py:1]: do we need this custom drift checker, could we drop this or e.g. replace with codegen if necessary

maybe some of these ideas would be better followed up with issues to resolve the underlying prototype codebase areas

File: web/e2e/build-site.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  // route, and a write that silently landed somewhere else. Without this a
+  // gap just renders a smaller dashboard, which every assertion below would
+  // happily pass.
+  //
+  // The two kinds of gap have different causes, so they are reported apart: a
+  // document is missing because `ROUTES` lacks it, while a chart or download is
+  // missing only if the emitting loops above have fallen out of step with what
+  // `referenced()` walks. One remedy pointed at both would send the reader to
+  // the wrong file half the time.
+  const missingDocuments = documents.filter((file) => !existsSync(join(API_DIR, file)));
```
Comment (exploreriii) [inline web/e2e/build-site.ts:111]: there is some dead code here

**Review by exploreriii — CHANGES_REQUESTED** [review]
Hi @Mounil2005 main feedback on this is are we able to streamline and simplify this, there is a lot of custom code being added here and shouldn't just wrap around problems already existingin the codebase
